synthesis.go

```diff
@@ -0,0 +1,262 @@
+do that for 1000 rounds, at which point we just give up and return the sequence to the user as is.
+
+Hopefully, this will enable more high throughput synthesis of genes!
+
+Keoni
+
+[started Dec 9, 2020]
+******************************************************************************/
+
+// DnaSuggestion is a suggestion of a fixer, generated by a problematicSequenceFunc.
+type DnaSuggestion struct {
+	Start          int    `db:"start"`
+	End            int    `db:"end"`
+	Bias           string `db:"gcbias"`
+	QuantityFixes  int    `db:"quantityfixes"`
+	SuggestionType string `db:"suggestiontype"`
+	Step           int    `db:"step"`
+	Id             int    `db:"id"`
+}
+
+// FindBsaI is a simple problematicSequenceFunc, for use in testing
+func FindBsaI(sequence string, c chan DnaSuggestion, wg *sync.WaitGroup) {
+	re := regexp.MustCompile(`GGTCTC`)
+	locs := re.FindAllStringIndex(sequence, -1)
+	for _, loc := range locs {
+		position := loc[0] / 3
+		leftover := loc[0] % 3
+		switch {
+		case leftover == 0:
+			c <- DnaSuggestion{position, (loc[1] / 3), "NA", 1, "BsaI removal", 0, 0}
+		case leftover != 0:
+			c <- DnaSuggestion{position, (loc[1] / 3) - 1, "NA", 1, "BsaI removal", 0, 0}
+		}
+	}
+	wg.Done()
+}
+
+// FindTypeIIS is a problematicSequenceFunc used for finding TypeIIS restriction enzymes. It finds BbsI, BsaI, BtgZI, BsmBI, SapI, and PaqCI(AarI)
+func FindTypeIIS(sequence string, c chan DnaSuggestion, wg *sync.WaitGroup) {
+	enzymeSites := []string{"GAAGAC", "GGTCTC", "GCGATG", "CGTCTC", "GCTCTTC", "CACCTGC"}
+	var enzymes []string
+	for _, enzyme := range enzymeSites {
+		enzymes = []string{enzyme, ReverseComplement(enzyme)}
+		for _, site := range enzymes {
+			re := regexp.MustCompile(site)
+			locs := re.FindAllStringIndex(sequence, -1)
+			for _, loc := range locs {
+				position := loc[0] / 3
+				leftover := loc[0] % 3
+				switch {
+				case leftover == 0:
+					c <- DnaSuggestion{position, (loc[1] / 3), "NA", 1, "TypeIIS removal", 0, 0}
+				case leftover != 0:
+					c <- DnaSuggestion{position, (loc[1] / 3) - 1, "NA", 1, "TypeIIS removal", 0, 0}
+				}
+			}
+		}
+	}
+	wg.Done()
+}
+
+func findProblems(sequence string, problematicSequenceFuncs []func(string, chan DnaSuggestion, *sync.WaitGroup)) []DnaSuggestion {
+	// Run functions to get suggestions
+	suggestions := make(chan DnaSuggestion, 100)
+	var wg sync.WaitGroup
+	for _, f := range problematicSequenceFuncs {
+		wg.Add(1)
+		go f(sequence, suggestions, &wg)
+	}
+	wg.Wait()
+	close(suggestions)
+
+	var suggestionsList []DnaSuggestion
+	for suggestion := range suggestions {
+		suggestionsList = append(suggestionsList, suggestion)
+	}
+	return suggestionsList
+}
+
+// FixCds fixes a CDS given the CDS sequence, a codon table, and a list of functions to solve for.
+func FixCds(sqlitePath string, sequence string, codontable CodonTable, problematicSequenceFuncs []func(string, chan DnaSuggestion, *sync.WaitGroup)) (string, error) {
```

Comment:
```suggestion
func generateKoeniHeuristicSequence(sqlitePath string, sequence string, codontable CodonTable, problematicSequenceFuncs []func(string, chan DnaSuggestion, *sync.WaitGroup)) (string, error) {
```
If this is a non standard generation technique, it needs more explanation. If it is a standard technique, it needs a link.

Reply:
I'm not really sure what you mean by a standard generation technique. This functions fixes CDSs so they can be synthesized. It's not for making heuristics or anything, it is specifically built so that users can send things to DNA synthesizers without having to worry about different sequence elements.

Reply:
> I'm not really sure what you mean by a standard generation technique

Like if I ran the sequence through a different library would you get the exact same result?

Becuase you are doing an estimate here: https://github.com/TimothyStiles/poly/pull/98/files/762d6766a445bfc287ed3e1014bc6ee854d8b961?file-filters%5B%5D=.go#diff-a7a2897e953e4b002cec604e45b7055c857a064bec21525d7b020d50538f1f4bR201
I assume not.

> It's not for making heuristics or anything
OK I see.

Wjhat guarantees do the user have about this sequence? Like is it within a certain error range? 

Or after 100 iterations are we ensured that we will get a result that is completely functional and all deviations are just noise?

Reply:
I assume this is the function that you expected to converge after a few times, but I'd still like to know how it deviates from other implementations and if those deviations are neglible to the resutl.

Reply:
TL;DR: no guarantees, but empirically good enough.

> Like if I ran the sequence through a different library would you get the exact same result?

Absolutely not. One thing about automated synthesis fixing is that it isn't really a popular thing right now. There is essentially 2 open source libraries that do this, and I wrote one of them (it kinda sucked). Funny enough, I actually did use Pandas in that one.

> What guarantees do the user have about this sequence? Like is it within a certain error range? Or after 100 iterations are we ensured that we will get a result that is completely functional and all deviations are just noise?

No real guarantees except I want to tell the user if we couldn't figure out a fix for their sequence.

> I'd still like to know how it deviates from other implementations

The only real competitor is https://edinburgh-genome-foundry.github.io/DnaChisel/ . The main difference is they use a constraint solving algorithm, while here we just directly fix things that we know are troublesome. 

I think the assumption you're coming with is that we know what we're looking for when it comes to synthesis fixing, but we really do not.

Do repeats of exactly 20 base pairs and no more and no less make it so you can't synthesize a given piece of DNA? Twist says they can't, Genscript says they can. Even if a synthesis provider says they can provide certain DNA molecules, they often can't (we experienced 20% failure rate with Twist with DNA they said they COULD make). When machine learning is applied to see what can or cannot be synthesized by given synthesis providers, we can approximate better results than the providers themselves. Sometimes, constraints on synthesis are also relaxed during times of excess production. 

For this implementation, I'm coming in with experience from FreeGenes of what Twist told me they can't do (though they could not do a lot more than just this). We synthesized about ~4-5 million base pairs there. As it turns out, if you do 3 things(1):

1. Remove repeats
2. Remove homopolymers
3. Remove high GC/low GC regions

While not allowing cyclic fixes (ie, same place getting flipped back and forth), you get ~5,225/5,228 genes to be successfully synthesis fixed (the other 3 you need small repeat testing). You don't need theoretically perfect methods for fixing most people's synthesis (like what DnaChisel)

(1) [Here is the implementation](https://github.com/Koeng101/dna_designer/blob/master/dna_designer/moclo.py) I used for FreeGenes. Don't judge me, it's academic code.

Reply:
Thanks for that detailed explanation. Very interesting and helpful.

I guess I would then suggest the prosaic `AttemptToFixCds` and copy some of this awesome comment into the func description.

Reply:
Added doc saying that we only attempt to make sequences good.